File: TPII-2025-2/PadraoGoF/Estrutural/Composite/Composite01.js
// Estrutura Componente:
class Componente{
    constructor(nome){
        this.nome = nome;
    }

    adicionar(){}

    remover(){}

    obterNome(){}
}

// Estrutura Folha:
class Folha extends Componente{
    constructor(nome, preco){
        super(nome);
        this.preco = preco;
    }

    obterNome(){
        return this.nome;
    }

    obterPreco(){
        return this.preco;
    }
}

// Estrutura Conteiner que irá compor (Componente):
class Conteiner extends Componente{
    constructor(nome){
        super(nome);
        this.componentes = [];
    }

    adicionar(componente){
        this.componentes.push(componente);
    }

    remover(componente){
        const index = this.componentes.indexOf(componente);
        this.componentes.splice(index, 1);
    }

    obterNome(){
        return this.nome;
    }

    obterPreco(){
        let preco = 0;
        this.componentes.forEach(componente => {
            preco += componente.obterPreco()
        });
        return preco;
    }
}

// Exemplo de Uso - Supermercado:
const maca = new Folha('Maçã', 2.00);
const laranja = new Folha('Laranja', 3.89);
const uva = new Folha('Uva', 4.99);

const frutas = new Conteiner('Frutas');
frutas.adicionar(maca);
frutas.adicionar(laranja);
frutas.adicionar(uva);

const caixa = new Conteiner('Compra Total');
caixa.adicionar(frutas);
caixa.adicionar(new Folha('Embalagem', 1.19));

console.log(`Preço Total: R$${caixa.obterPreco()}`);

File: TPII-2025-2/PadraoGoF/Estrutural/Composite/Composite01.py
# Estrutura Componente:
class Componente:
    def __init__(self, nome):
        self.nome = nome

    def adicionar(self):
        pass

    def remover(self):
        pass

    def obter_nome(self):
        pass

# Estrutura Folha:
class Folha(Componente):
    def __init__(self, nome, preco):
        super().__init__(nome)
        self.preco = preco

    def obter_nome(self):
        return self.nome

    def obter_preco(self):
        return self.preco

# Estrutura Conteiner que irá compor (Componente):
class Conteiner(Componente):
    def __init__(self, nome):
        super().__init__(nome)
        self.componentes = []

    def adicionar(self, componente):
        self.componentes.append(componente)

    def remover(self, componente):
        if componente in self.componentes:
            self.componentes.remove(componente)

    def obter_nome(self):
        return self.nome

    def obter_preco(self):
        preco = 0
        for componente in self.componentes:
            preco += componente.obter_preco()
        return preco

# Exemplo de Uso - Supermercado:
maca = Folha('Maçã', 2.00)
laranja = Folha('Laranja', 3.00)
uva = Folha('Uva', 5.00)

frutas = Conteiner('Frutas')
frutas.adicionar(maca)
frutas.adicionar(laranja)
frutas.adicionar(uva)

caixa = Conteiner('Compra Total')
caixa.adicionar(frutas)
caixa.adicionar(Folha('Embalagem', 2.00))

print(f'Preço Total: R${caixa.obter_preco():.2f}')
